### ISSUE TITLE
Keep service stopped during and after deploy

### ISSUE THREAD

**Issue by pre** [kontena/kontena#2865]
I have a stack with a bunch of worker services. I have strong reasons not to run any workers during and after the deploy of new code, until all pending database migrations have been executed.

This is what I have now:

1) service stop
2) service deploy
3) run database migrations (by `rake db:migrate` using one of the worker services)
4) service start

Problem is that `service deploy` starts the services even though they were stopped before the deploy. This results in the new code being executed although the database does not yet contain the changes which are assumed by the new code. If I stop services immediately after the deploy, there is still a window when new code might get executed before the new stop command is in effect. 

Is there any way to keep services stopped during and after the deploy? 

I am currently falling back to using an environment variable like `KEEP_SERVICES_STOPPED` which prevents worker services from starting (eg. loop with `sleep 9999999`  if it's set).

The database is not stopped at any point of time. I only want to prevent the worker services from running.

**Comment by SpComb** [comment]
I assume that in step 2. you mean `kontena stack deploy`? With the db and workers being separate services, I assume you'd specifically want to deploy the db service, run the migrations, and then deploy (and start) the worker services. However, due to the way that `kontena stack upgrade` currently works (#2311), you can't use the `--no-deploy` flag to control the service deploys manually.

How exactly are you running the database migrations in step 3? Could you run them as, say, post_start hooks on the database service / the new 1.4 `pre_start` hooks on the worker service? During the stack -> service deploys, kontena will wait for the service hooks to execute before continuing with the rest of the deploy.

Although, currently, `post_start` hook errors won't stop the rest of the deploy from continuing (#2843), so if the migrations fail, the workers will still get started.

**Comment by pre** [comment]
@SpComb I think you misunderstood me. I mean service deploy. The database is not even running in Kontena.

Database migrations come from the application which owns the database. The deploy script goes currently like this:

```bash
#!/usr/bin/env bash

set -e

SIDEKIQ_WORKER="stackname/lolcats"
SNEAKERS_WORKER="stackname/longcats"

function service_stop { kontena service stop "$@"; }
function service_start { kontena service start "$@"; }
function service_exec { kontena service exec "$@"; }
function service_deploy { kontena service deploy "$@"; }

bin/docker-production-build

service_stop ${SIDEKIQ_WORKER}
service_stop ${SNEAKERS_WORKER}

service_deploy ${SIDEKIQ_WORKER}
service_deploy ${SNEAKERS_WORKER}

service_exec ${SIDEKIQ_WORKER} rake db:migrate

service_start ${SIDEKIQ_WORKER}
service_start ${SNEAKERS_WORKER}
```

**Comment by pre** [comment]
I clarified my original question, that the database is never stopped. I only want to make sure that workers are not executed until the database migrations are in.

I'm looking for a "maintenance mode" kind of solution which would temporarily disable a service.

**Comment by kke** [comment]
A workaround could be to have something in the worker initializer that sleeps while there are pending migrations. But I don't think a stopped service should be started when upgraded.

**Comment by SpComb** [comment]
You won't be able to `kontena service exec` if the service container is stopped, so just leaving the service stopped during the deploy wouldn't allow for that kind of usecase.

Your example looks like it would be a good usecase for the new `pre_start` hooks that are coming in 1.4. Those run as a separate container before the service container itself is started, and then you can use `oneshot: true` naming trickery to control when they get run. Renaming the hook on every version bump would allow you to control when the migrations get run:

```
hooks:
  pre_start:
    - name: db migrate v1
      cmd: rake db:migrate
      instances: 1
      oneshot: true
``` 

Workaround would be to include the `rake db:migrate` in the actual service container command itself, and either run it every time, or have some logic of your own to determine when it should run.

**Comment by pre** [comment]
It seems safest to set an env variable `MAINTENANCE_MODE` for the duration of the deploy, and then loop with `sleep 9999999` in the worker's startup initializer (`sidekiq.rb` and `sneakers.rb`). This would prevent the application level code from starting while it allows `kontena service exec rake db:migrate`.

It sounds risky to run `rake db:migrate` in the container's startup command, as there are many instances running in parallel.

Any option where the stack or `kontena.yml` needs to be updated are out of the question since they are quite incompatible with automated deployes where only the service needs to be deployed. Stack upgrade would unnecessary restart all the other services in that stack and it sounds awkward to let eg. CI to modify kontena.yml on the fly.

**Comment by SpComb** [comment]
If you need to run the migrations separately from the application, then trying to run the migrations as part of the application container is going to be a little awkward. It sounds like the ideal solution for you would be some kind of separate "task" support: some kind of one-off services that Kontena does not normally expect to be running, but lets you to manually trigger them instead, and Kontena just waits for them to finish without restarting them. Kind of like execs, but not needing a dummy container just running a `sleep` loop.

The new `pre_start` hooks in 1.4 are already kind of like that, they get run as one-off containers and Kontena waits for them to exit before spawning the actual service container. But that still needs *something* to decide when to actually run those migrations. If you want to use the `pre_start` hooks for automated schema upgrades, then you need to express that logic in a declarative way: either encode some kind of `with_lock { upgrade if needs_upgrade? }` logic into the actual hook itself and run them always (like the `kontena-server` app on the Kontena master nodes itself does!), or use the Kontena `oneshot: true` + `instances: 1` hook logic with some kind of version numbering to control exactly when they get run.

For any kind of non-declarative workflow where you manually decide when to `kontena service stop` -> run migrations -> `kontena service deploy`, you would need some kind of "task" support. The current workaround for running manual tasks is to have a dummy container that just sleeps doing nothing, and then run execs on it. So instead of the `MAINTENANCE_MODE`, you would just have a separate "migration" service that just always sleeps. Then for upgrades, you stop everything, deploy the "migration" service with the new code, exec the migration on it, and then deploy the app services to start running the new code.

> It sounds risky to run rake db:migrate in the container's startup command, as there are many instances running in parallel.

You can use `instances: 1` in the hooks to only run it on the first instance. If you link to the migrating service from the other services, and only deploy the migrating service one instance at a time (either `min_health: 1.0` or keep it to a single instance), you can guarantee that the migration hooks will have run before any other services get deployed.

> Stack upgrade would unnecessary restart all the other services in that stack

It should only re-deploy/restart services that have actually changed, barring some exceptions related to Kontena version upgrades that trigger service updates. If `kontena stack upgrade` always re-creates/restarts the stack services, then that's a bug. We fixed some of those in 1.2 (#2090).

> Any option where the stack or kontena.yml needs to be updated are out of the question since they are quite incompatible with automated deployes where only the service needs to be deployed.

Do you mean you deploy new versions of the application without upgrading the stack? I guess that means that you're using `image: foo:latest`, and then using `kontena service deploy` to trigger an image pull and have the containers get re-created if the image has changed?

The other side of that is to have a version number for each deploy, and upgrade the stack with the service images pinned to that specific version.

> it sounds awkward to let eg. CI to modify kontena.yml on the fly.

You can use stack variables to interpolate the hook name, and have the CI pass the migration version from somewhere else.

**Comment by pre** [comment]
Thank you for the thoughtful answer. I will consider the choices in the upcoming weeks. At the moment we're on Kontena 1.3, I'll ask if we could upgrade.


 Petrus
